master/buildbot/scripts/devproxy.py

```diff
@@ -0,0 +1,168 @@
+        if plugins is None:
+            plugins = {}
+        else:
+            plugins = json.loads(plugins)
+
+        self.plugins = plugins
+
+        app.router.add_route('*', '/ws', self.ws_handler)
+        for path in ['api', 'auth', 'sse', 'avatar']:
+            app.router.add_route(
+                '*', '/%s{path:.*}' % path, self.proxy_handler)
+        app.router.add_route('*', '/', self.index_handler)
+        for plugin in self.apps.names:
+            if plugin != 'base':
+                staticdir = self.apps.get(plugin).static_dir
+                app.router.add_static('/' + plugin, staticdir)
+        staticdir = self.staticdir = self.apps.get('base').static_dir
+        loader = jinja2.FileSystemLoader(staticdir)
+        self.jinja = jinja2.Environment(
+            loader=loader, undefined=jinja2.StrictUndefined)
+        app.router.add_static('/', staticdir)
+        conn = aiohttp.TCPConnector(
+            limit=self.MAX_CONNECTIONS, verify_ssl=(not self.unsafe_ssl))
+        self.session = aiohttp.ClientSession(connector=conn, trust_env=True, cookies=cookies)
+        self.config = None
+        self.buildbotURL = "http://localhost:{}/".format(port)
+        aiohttp.web.run_app(app, host="localhost", port=port)
+        self.session.close()
+
+    async def ws_handler(self, req):
+        # based on https://github.com/oetiker/aio-reverse-proxy/blob/master/paraview-proxy.py
+        ws_server = aiohttp.web.WebSocketResponse()
+        await ws_server.prepare(req)
+
+        async with self.session.ws_connect(
+            self.next_url + "/ws", headers=req.headers
+        ) as ws_client:
+
+            async def ws_forward(ws_from, ws_to):
+                async for msg in ws_from:
+                    mt = msg.type
+                    md = msg.data
+                    if mt == aiohttp.WSMsgType.TEXT:
+                        await ws_to.send_str(md)
+                    elif mt == aiohttp.WSMsgType.BINARY:
+                        await ws_to.send_bytes(md)
+                    elif mt == aiohttp.WSMsgType.PING:
+                        await ws_to.ping()
+                    elif mt == aiohttp.WSMsgType.PONG:
+                        await ws_to.pong()
+                    elif ws_to.closed:
+                        await ws_to.close(code=ws_to.close_code, message=msg.extra)
+                    else:
+                        raise ValueError('unexpected message type: %s' % msg)
+
+            # keep forwarding websocket data in both directions
+            await asyncio.wait([ws_forward(ws_server, ws_client), ws_forward(ws_client, ws_server)], return_when=asyncio.FIRST_COMPLETED)
+        return ws_server
+
+    async def proxy_handler(self, req):
+        method = getattr(self.session, req.method.lower())
+        upstream_url = self.next_url + req.path
+        headers = req.headers.copy()
+        query = req.query
+        try:
+            async with method(
+                    upstream_url,
```

Comment:
Inconsistent indentation.

Reply:
Interesting. Maybe this could work:
```
async with method(upstream_url,
                  other_arg,
                  another_arg) as request:
```

Reply:
it does. I changed  the code to use this method